ops.qmd

```diff
@@ -2,84 +2,760 @@
+
+Establishing shared data contexts - glossaries, [data dictionaries](https://en.wikipedia.org/wiki/Data_dictionary), schema references - ensures alignment on data meaning and usage across roles. Documentation aids understanding for those without direct data access.
+
+For example, a data scientist may use Weights & Biases to analyze an anomaly detection model experiment and share the evaluation results with other team members to discuss improvements. The final model can then be registered with MLflow before handing off for deployment.
+
+Enabling transparency, traceability and communication via MLOps empowers teams to remove bottlenecks and accelerate delivery of impactful ML systems.
+
+
+## Hidden Technical Debt in ML Systems
+
+Technical debt is an increasingly pressing issue for ML systems. This metaphor, originally proposed in the 1990s, likens the long-term costs of quick software development to financial debt. Just as some financial debt powers beneficial growth, carefully managed technical debt enables rapid iteration. However, left unchecked, accumulating technical debt can outweigh any gains.
+
+![Components of a production ML system with "ML code" highlighted as a small portion of system](images/ai_ops/hidden_debt.png)
+*Components of a production ML system - the ML code is only a very small portion of the system*
+
+### Model Boundary Erosion 
+Unlike traditional software, ML lacks clear boundaries between components as seen in the diagram above. This erosion of abstraction creates entanglements that exacerbate technical debt in several ways:
+
+### Entanglement
+
+Tight coupling between ML model components makes isolating changes difficult. Modifying one part causes unpredictable ripple effects throughout the system. Changing anything changes everything (also known as CACE) is a phenomenon that applies to any tweak you make to your system. Potential mitigations include decomposing the problem when possible or closely monitoring for changes in behavior to contain their impact.
+
+### Correction Cascades
+
+![Diagram showing some tasks affect other future tasks](images/ai_ops/data_cascades.png)
+
+Building models sequentially creates risky dependencies where later models rely on earlier ones. For example, taking an existing model and fine-tuning it for a new use case seems efficient. However, this bakes in assumptions from the original model that may eventually need correction. Modifying foundational components then becomes extremely costly due to the cascading effects on subsequent models. One mitigation is to only augment existing models when absolutely necessary to reuse some capabilities. Often, it is safer to train models for new use cases from scratch to avoid baking in dependencies. Careful thought should be given to identifying points where introducing fresh model architectures can avoid correction cascades down the line.
```

Comment:
I wonder if it'd be worth explicitly defining what a "correction cascade" is in this paragraph?

Reply:
Done, thanks for the feedback.